Check in public for matching webpackShims

w33ble opened the pull request:
Closes #13

Check for a matching `webpackShims` path, starting at the plugin root, walk all the way to the requesting file, then check Kibana's root. Stops as soon as it finds a match, so in most cases, only the first condition is needed.

w33ble committed:
check in public webpackShims

also simlify the webpackShim checking a bit by using reduce

spalger commented on index.js:
```diff
@@ -119,22 +119,25 @@ function resolveLocalRelativeImport(fileImport, file) {
+  const baseSource = sourceParts.pop();
+
+  const pluginShimPaths = [
+    path.join(rootPath, 'webpackShims', ...sourceParts),
```
The `webpackShims` directory could be in any directory relative to the file up to the root, so these paths need to include checks for intermediate checks. Maybe something like this?

```js
const relativeWebpackShims = sourceParts.reduce((acc, sourcePart) => [
  ...acc,
  path.join(acc[acc.length - 1], sourcePart, 'webpackShims')
], [
  path.join(rootPath, 'webpackShims')
])

const pluginShimPaths = [
  ...relativeWebpackShims,
  path.join(kibanaPath, 'webpackShims', ...sourceParts)
]
```

spalger commented on index.js:
```diff
@@ -119,22 +119,25 @@ function resolveLocalRelativeImport(fileImport, file) {
+  const baseSource = sourceParts.pop();
+
+  const pluginShimPaths = [
+    path.join(rootPath, 'webpackShims', ...sourceParts),
```
The `webpackShims` directory could be in any directory relative to the file up to the root, so this needs to check for `webpackShims` directories starting at the basedir of the requested file up to the root. The console plugin, for instance, has it's own `webpackShims` directory, and any directory can have a `webpackShims` directory just like it can have a `node_modules` directory.

w33ble replied:
> The webpackShims directory could be in any directory relative to the file up to the root

I did not know that. That makes resolution a bit trickier. I'm betting there code in Kibana that we can borrow here for the resolution. In any event, this PR will need to be updated then.

w33ble committed:
fix path separator issue

w33ble committed:
recurse from root for webpackShims

check for a matching webpackShims path, starting at the plugin root, walk all the way to the requesting file, then check kibana's root

spalger commented:
Superseded by #15
